### DISCUSSION TITLE
FPP Math Component Tutorial Generate Step Failure

### DISCUSSION THREAD

**Discussion by astroesteban** [nasa/fprime#1235] [Q&A]
| | |
|:---|:---|
|**_F´ Version_**| v3.0.0 and devel|
|**_Affected Component_**| FPP  |
---

System information: I am running F' inside of an Ubuntu 20.04 Docker image.

## Problem Description

I was going through the Math Component Tutorial to make some updates and in section **2.3. Build the Model** I got the error when I ran `prime-util generate`:
```sh
[WARNING] Failed to find settings file: /workspaces/fprime/Ref/settings.ini
[INFO] Generating build directory at: /workspaces/fprime/Ref/build-fprime-automatic-native
[INFO] Using toolchain file None for platform default
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Searching for F prime modules in: /workspaces/fprime
-- Autocoder constants file: /workspaces/fprime/config/AcConstants.ini
-- Configuration header directory: /workspaces/fprime/config
-- [fpp-tools] Searching for fpp-tools
-- [fpp-tools] Searching for sbt
-- [fpp-tools] sbt found at: /usr/bin/sbt
CMake Error at /workspaces/fprime/cmake/fpp-download/fpp.cmake:81 (message):
  Build step for fpp failed: 2

  Cloning into 'src'...

  Note: switching to 'ee355fc99eb8040157c62e69f58ac6a8435cd981'.

  

  You are in 'detached HEAD' state.  You can look around, make experimental

  changes and commit them, and you can discard any commits you make in this

  state without impacting any branches by switching back to a branch.

  

  If you want to create a new branch to retain commits you create, you may

  do so (now or later) by using -c with the switch command.  Example:

  

    git switch -c <new-branch-name>

  

  Or undo this operation with:

  

    git switch -

  

  Turn off this advice by setting config variable advice.detachedHead to
  false

  

  HEAD is now at ee355fc9 Fix typo in User's Guide

  make[2]: *** [CMakeFiles/fpp-tools.dir/build.make:75:
  fpp-tools-prefix/src/fpp-tools-stamp/fpp-tools-install] Error 2

  make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/fpp-tools.dir/all] Error
-- [fpp-tools] Installing new tools to /workspaces/fprime/Ref/build-fprime-automatic-native/fpp-tools-install, this will take a few minutes
-- Configuring incomplete, errors occurred!
  2
See also "/workspaces/fprime/Ref/build-fprime-automatic-native/CMakeFiles/CMakeOutput.log".

  make: *** [Makefile:84: all] Error 2

Call Stack (most recent call first):
  /workspaces/fprime/cmake/required.cmake:20 (include)
  /workspaces/fprime/cmake/FPrime.cmake:26 (include)
  CMakeLists.txt:32 (include)


[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

I didn't see any other open issues for this but I did follow the instructions in #1205.

I proceeded to install  SBT in my Docker image:
```docker
RUN echo "deb https://repo.scala-sbt.org/scalasbt/debian all main" | tee /etc/apt/sources.list.d/sbt.list \
    && echo "deb https://repo.scala-sbt.org/scalasbt/debian /" | tee /etc/apt/sources.list.d/sbt_old.list \
    && curl -sL "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x2EE0EA64E40A89B84B2DF73499E82A75642AC823" | apt-key add \
    && apt-get update \
    && apt-get install sbt
```

and then manually installed FPP from `main`.

After that I was able to successfully run `fprime-utils generate`

Is this a known issue? Should we expand upon the installation instructions to mention SBT and FPP setup?

**Comment by LeStarch** [comment]
I think this is the first time we have seen this.  We are working on performance with FPP and with that will come a simplified setup.  Our intention is to publish binary executables that can just be added to the PATH.

**Reply by astroesteban** [reply]
Awesome! Good to know! Thanks @LeStarch!

**Reply by astroesteban** [reply]
FYI: I cloned and installed fpp `v1.0.0` and I kept getting the warning:
```cmake
/usr/local/bin/fpp/fpp-depend version v1.0.0 commit ee355fc99eb8040157c62e69f58ac6a8435cd981 not expected version 1f3fdbfdd1832c9ec9b2acae6de3ba45669d4851
```

Then the fpp.cmake module would proceed to download fpp and finally error out with the error I posted above.

The only way I got it to work is if I checked out hash `1f3fdbfdd1832c9ec9b2acae6de3ba45669d4851` which is what is hardcoded in fpp.cmake:

```docker
# install sbt for fpp
RUN echo "deb https://repo.scala-sbt.org/scalasbt/debian all main" | tee /etc/apt/sources.list.d/sbt.list \
    && echo "deb https://repo.scala-sbt.org/scalasbt/debian /" | tee /etc/apt/sources.list.d/sbt_old.list \
    && curl -sL "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x2EE0EA64E40A89B84B2DF73499E82A75642AC823" | apt-key add \
    && apt-get update \
    && apt-get install sbt

# install fpp
WORKDIR /workspace/fpp
RUN git clone https://github.com/fprime-community/fpp.git . \
    && cd compiler \
    && git checkout 1f3fdbfdd1832c9ec9b2acae6de3ba45669d4851 \
    && export FPP_SBT_FLAGS='--batch -Dsbt.server.forcestart=true' \
    && ./install /usr/local/bin/fpp

ENV FPP_INSTALL_DIR=/usr/local/bin/fpp
ENV PATH=${PATH}:${FPP_INSTALL_DIR}
```

Should F' _require_ that specific version of FPP?

**Reply by LeStarch** [reply]
At this time, the FPP tools are not guaranteed to be backwards compatible.  Thus for a given F´ version, a specific FPP version is required to guarantee all features work.

We intend to move toward the following system:

1. Tagged versions of F´  pair with a tagged version of FPP that includes native installations for Linux/MacOs.  Thus, using a tagged version of F´ will be easy to setup and install.
2. Automatic installations of tools will be discontinued.  FPP will check versions, but not install on failure.  CMake best practices cautions against pulling in executable onto users' computers
3. Developers should checkout and build the specified version of FPP as devel will continue to track a specific version.
4. An override flag will be added for developers who know their version of tools work, to bypass that check.

**Reply by LeStarch** [reply]
I am curious to hear if those points meet your expectations, or if we need to reconsider.

**Reply by LeStarch** [reply]
Also, we are thinking to change the default branch to `master`.  That way the FPP tools version is not a moving target.  If you work with `devel` you may need to install specific tools versions....but if you work with the default checkout, changes will be slower, as will needs to update tools.

We'd put a requirement that merge-to-master come with a native set of tools you can install.

**Reply by astroesteban** [reply]
In response to your first point. I like that idea. If semantic versioning is enforced for FPP could a range of versions be supported in F'? Instead of tying F' to a very specific MAJOR.MINOR.PATCH could it be tied to a MAJOR or MAJOR.MINOR? That way we get the latest bug fix for the specific FPP version. It also helps remove the overhead of having to update the hardcoded commit hash every time there's an update to FPP that F' needs.

This is similar to what you can do in setup.py with dependencies:
`numpy ~= 3.4` will tie you to `v3.4` but you can get the latest `3.4.xx` update.

The rest of your points make perfect sense to me. Moving that out of the CMake configuration will help speed things up and make things more secure.

**Comment by ThibFrgsGmz** [comment]
Just to let you know that I also have the same problem now. I'm going to analyze your discussion to get unstuck.

```cmake
See also "/mnt/c/Users/ThibFrgsGmz/fprime_lab/fprime/Ref/build-fprime-automatic-native/CMakeFiles/CMakeOutput.log".

  Cloning into 'src'...

  Note: switching to 'v1.0.1'.



  You are in 'detached HEAD' state.  You can look around, make experimental

  changes and commit them, and you can discard any commits you make in this

  state without impacting any branches by switching back to a branch.



  If you want to create a new branch to retain commits you create, you may

  do so (now or later) by using -c with the switch command.  Example:



    git switch -c <new-branch-name>



  Or undo this operation with:



    git switch -



  Turn off this advice by setting config variable advice.detachedHead to
  false



  HEAD is now at a8705b3e Update version number in docs

  make[2]: *** [CMakeFiles/fpp-tools.dir/build.make:75:
  fpp-tools-prefix/src/fpp-tools-stamp/fpp-tools-install] Error 2

  make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/fpp-tools.dir/all] Error
  2

  make: *** [Makefile:84: all] Error 2

Call Stack (most recent call first):
  /mnt/c/Users/ThibFrgsGmz/fprime_lab/fprime/cmake/required.cmake:21 (include)
  /mnt/c/Users/ThibFrgsGmz/fprime_lab/fprime/cmake/FPrime.cmake:25 (include)
  CMakeLists.txt:32 (include)


[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

**Reply by astroesteban** [reply]
I manually installed FPP and it worked. They key part is that the commit you checkout to in FPP matches what's in fpp.cmake. In this case v1.0.1. That way F Prime doesn't try to clone FPP.

Here's that section from the Dockerfile I used:
```docker
# install sbt for fpp
RUN echo "deb https://repo.scala-sbt.org/scalasbt/debian all main" | tee /etc/apt/sources.list.d/sbt.list \
    && echo "deb https://repo.scala-sbt.org/scalasbt/debian /" | tee /etc/apt/sources.list.d/sbt_old.list \
    && curl -sL "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x2EE0EA64E40A89B84B2DF73499E82A75642AC823" | apt-key add \
    && apt-get update \
    && apt-get install sbt

# install fpp
WORKDIR /workspace/fpp
RUN git clone https://github.com/fprime-community/fpp.git . \
    && git checkout v1.0.1 \
    && cd compiler \
    && ./install /usr/local/bin/fpp
ENV FPP_INSTALL_DIR=/usr/local/bin/fpp
ENV PATH=${PATH}:${FPP_INSTALL_DIR}
```

**Reply by ThibFrgsGmz** [reply]
It worked!

**Reply by ThibFrgsGmz** [reply]
I don't know if you tested it on WSL-2 but I always get this error when I restart my computer (not sure if it's true because I'm doing a lot of things in parallel so I'm not really focused).

I have to redefine the path to the FPP tools every time because it is not perennial in the PATH. At least in my configuration.

**Reply by ThibFrgsGmz** [reply]
OK this is because I didn't add the export command to my ``~/.bashrc`` file.
